### ISSUE TITLE
Random "Can't perform a React state update on an unmounted component..." errors in v0.2.0.

### ISSUE THREAD

**Issue by chybisov** [facebookexperimental/Recoil#951]
After upgrading to version v0.2.0 I'm getting errors like this at random components and can't figure out why it happens. All components works fine with v0.1.2. All my atoms just simple atoms without async selectors and other complicated stuff.

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
at node_modules\react-native\Libraries\LogBox\LogBox.js:173:8 in registerError
at node_modules\react-native\Libraries\LogBox\LogBox.js:59:8 in errorImpl
at node_modules\react-native\Libraries\LogBox\LogBox.js:33:4 in console.error
at node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:106:4 in printWarning
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:75:16 in error
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:19152:9 in warnAboutUpdateOnUnmountedFiberInDEV
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:17093:40 in scheduleUpdateOnFiber
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:11003:16 in dispatchAction
at [native code]:null in dispatchAction
at node_modules\recoil\native\recoil.js:4309:24 in subscribeToRecoilValue$1$argument_2
at node_modules\recoil\native\recoil.js:3321:41 in sendEndOfBatchNotifications
at node_modules\recoil\native\recoil.js:3374:6 in Recoil_Queue.enqueueExecution$argument_1
at node_modules\recoil\native\recoil.js:1041:0 in enqueueExecution
at node_modules\recoil\native\recoil.js:3355:18 in useEffect$argument_0
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:15561:31 in commitHookEffectListMount
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:15618:35 in commitPassiveHookEffects
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:265:4 in invokeGuardedCallbackImpl
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:476:2 in invokeGuardedCallback
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:18795:29 in flushPassiveEffectsImpl
at node_modules\scheduler\cjs\scheduler.development.js:653:23 in unstable_runWithPriority
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:17650:21 in performSyncWorkOnRoot
at [native code]:null in performSyncWorkOnRoot
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:5321:31 in runWithPriority$argument_1
at node_modules\scheduler\cjs\scheduler.development.js:653:23 in unstable_runWithPriority
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:5316:21 in flushSyncCallbackQueueImpl
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:5304:28 in flushSyncCallbackQueue
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:17125:30 in scheduleUpdateOnFiber
at node_modules\react-native\Libraries\Renderer\implementations\ReactNativeRenderer-dev.js:11003:16 in dispatchAction
at [native code]:null in dispatchAction
at node_modules\recoil\native\recoil.js:3354:2 in Batcher
at node_modules\recoil\native\recoil.js:3556:4 in replaceState
at node_modules\recoil\native\recoil.js:775:9 in applyActionsToStore
at node_modules\recoil\native\recoil.js:798:9 in queueOrPerformStateUpdate
at node_modules\recoil\native\recoil.js:841:24 in setRecoilValue
at node_modules\recoil\native\recoil.js:4416:9 in useSetRecoilState
at node_modules\expo-image-manipulator\build\ImageManipulator.js:4:7 in manipulateAsync
at node_modules\regenerator-runtime\runtime.js:63:36 in tryCatch
at node_modules\regenerator-runtime\runtime.js:293:29 in invoke
at node_modules\regenerator-runtime\runtime.js:63:36 in tryCatch
at node_modules\regenerator-runtime\runtime.js:154:27 in invoke
at node_modules\regenerator-runtime\runtime.js:164:18 in PromiseImpl.resolve.then$argument_0
at node_modules\react-native\node_modules\promise\setimmediate\core.js:37:13 in tryCallOne
at node_modules\react-native\node_modules\promise\setimmediate\core.js:123:24 in setImmediate$argument_0
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:130:14 in _callTimer
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:181:14 in _callImmediatesPass
at node_modules\react-native\Libraries\Core\Timers\JSTimers.js:441:30 in callImmediates
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:387:6 in __callImmediates
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:135:6 in __guard$argument_0
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:364:10 in __guard
at node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:134:4 in flushedQueue
at [native code]:null in flushedQueue
at [native code]:null in invokeCallbackAndReturnFlushedQueue
```

### Environment

| software                     | version
| ---------------------------- | -------
| expo                         | 40.0.1
| react                 | 17.0.2
| react-native                 | 0.63.4
| recoil                         | 0.2.0

**Comment by mondaychen** [comment]
Can you provide your code for reproduction? You can fork this codesandbox if that works https://codesandbox.io/s/recoil-basic-example-wti09?file=/src/App.js

**Comment by chybisov** [comment]
> Can you provide your code for reproduction? You can fork this codesandbox if that works https://codesandbox.io/s/recoil-basic-example-wti09?file=/src/App.js

Unfortunately no, because I don't know which part of code causing this issue.

**Comment by chybisov** [comment]
@mondaychen actually if you look on stack trace there is no mention of component name that have this issue, instead it only has mentions of recoil.

**Comment by cyphire** [comment]
I have similar issues.... It's not just Recoil but react native which has no way of knowing where errors come from.... Please let me know if you find a solution!

**Comment by cyphire** [comment]
Hey - I see that expo-image-manipulator is in your stack trace.... Are you using this module in recoil? If you comment out the part of your program which uses it does it eliminate the errors?

expo-image-manipulator calls async stuff.  If your module updates something in recoil, but the component no longer exists (it was rerendered) and it trys to update one of your atoms (with a set), this is probably why it is failing...

**Comment by chybisov** [comment]
@cyphire I use `expo-image-manipulator` on some screens, but components that uses it not even rendered at the time of an error. I rarely see random warnings `Warning: Can't perform a React state update on an unmounted component.` with `expo-image-manipulator` in trace even with Recoil v0.1.2, but in components where I'm not using Recoil. Could you please share your findings or do you know how to debug this issue?

**Comment by chybisov** [comment]
@cyphire just checked and removing `expo-image-manipulator` do nothing. I notice that it always happens when some component unmount, could it be some kind of issue with Recoil itself when it can't correctly unsubscribe component from updates?

@drarmstr sorry for bothering you, but could you please advice something? Thank you!

**Comment by chybisov** [comment]
@cyphire @drarmstr I just figure out the issue. I have a couple of places with code similar to this one

```
const useSomeHook = () => {
  const [state, setState] = useRecoilState(someAtom);
  useEffect(() => {
    setState((state) => ({
      ...state,
    }));
  }, [setState]);
  useEffect(() => {
    setState((state) => ({
      ...state,
    }));
  }, [setState]);
};
```

If you go to a page with this hook ones, then leave it, and come back again it produce that warning.

**Comment by Calneideck** [comment]
Just want to add, this is also happening to us in React (so not just react-native). Can't provide any code as the error is too generic but it appears to be related to components unmounting.

**Comment by mondaychen** [comment]
@chybisov I'm not able to reproduce it with this code. Do you have something else, maybe an async call in your `useEffect` callback?
Also I'd avoid setting recoil states in `useEffect` callback if possible.

**Comment by chybisov** [comment]
@mondaychen did you try with react native? It works for me every time. I use react native with expo if that matters.

Could you please explain why it's a bad idea to set Recoil state in useEffect?

**Comment by cyphire** [comment]
```const useSomeHook = () => {
  const [state, setState] = useRecoilState(someAtom);
  useEffect(() => {
    setState((state) => ({
      ...state,
    }));
  }, [setState]);
  useEffect(() => {
    setState((state) => ({
      ...state,
    }));
  }, [setState]);
};
```

Does not seem correct.

First of all the [setState] should be [state], not setState.  setState is the function which sets the state, the state itself you call 'state'.  useEffect is looking for an array of values which might change and when they change the useEffect gets called.

The setState should just be setState(obj) or whatever you are setting the state to....

It's not redux, you don't pass a function in....

```
const useSomeHook = (props) => {
  
  const [state, setState] = useRecoilState(someAtom);

  let {
     obj
  } = props;

  useEffect(() => {
    if (obj != null) {
      setState(obj);
    }
  }, [state]);

```

**Comment by cyphire** [comment]
I'm having a similar issue with the unmounted component... What is happening is that the useEffect is being called, but you have already destroyed that component as it is re-rendering and being recreated.

There is a lot of stuff out there about how to trap for this, and it's a major problem that I have with React / React-Native.... I'm about to fix this myself where i have this problem, you are going to need to drop in a useRef, see if the component still exists before any states are set...

So just to reiterate... anytime you call the setter function (setState() in your case), you need to see if the component still exists!

**Comment by jj-143** [comment]
```jsx
const stepState = atom({
  key: "step",
  default: "one",
})

function Example() {
  const step = useRecoilValue(stepState)
  return <>{step === "one" ? <One /> : step === "two" ? <Two /> : null}</>
}

function One() {
  const [step, setStep] = useRecoilState(stepState)
  // const setStep = useSetRecoilState(stepState)
  return (
    <>
      <h1>one</h1>
      <button onClick={() => setStep("two")}>next</button>
    </>
  )
}
function Two() {
  const [step, setStep] = useRecoilState(stepState)
  // const setStep = useSetRecoilState(stepState)
  return (
    <>
      <h1>two</h1>
      <button onClick={() => setStep("one")}>prev</button>
    </>
  )
}
```
This reproduces the warning in React, and checked v0.1.3 doesn't.
Using `useSetRecoilState` instead of `useRecoilState` prevents the warning (obviously?).

One thing I noticed is, the first step change (`one` -> `two`) makes no warning, but the second one (`two -> one`) shows the warning and the third (`one` -> `two`), and no more warning after.

**Comment by mondaychen** [comment]
@jj-143 I pasted your code into a codesandbox but didn't see any warning in console. Anything I missed? https://codesandbox.io/s/recoil-update-in-unmounted-component-example-81y7o?file=/src/App.js

**Comment by jj-143** [comment]
@mondaychen oh, I just tested with latest react, react-dom @17.0.2 with the same code and no warning, sorry about that. The version i reproduced with was @16.14.0.
https://codesandbox.io/s/recoil-update-in-unmounted-component-example-forked-o0rqr?file=/src/index.js
Maybe something changed after 16.14.0 which fixes my particular example, but can't say the same to the original issuer's code.

**Comment by chybisov** [comment]
> ```
>   const [state, setState] = useRecoilState(someAtom);
>   useEffect(() => {
>     setState((state) => ({
>       ...state,
>     }));
>   }, [setState]);
>   useEffect(() => {
>     setState((state) => ({
>       ...state,
>     }));
>   }, [setState]);
> };
> ```
> 
> Does not seem correct.
> 
> First of all the [setState] should be [state], not setState. setState is the function which sets the state, the state itself you call 'state'. useEffect is looking for an array of values which might change and when they change the useEffect gets called.
> 
> The setState should just be setState(obj) or whatever you are setting the state to....
> 
> It's not redux, you don't pass a function in....
> 
> ```
> const useSomeHook = (props) => {
>   
>   const [state, setState] = useRecoilState(someAtom);
> 
>   let {
>      obj
>   } = props;
> 
>   useEffect(() => {
>     if (obj != null) {
>       setState(obj);
>     }
>   }, [state]);
> ```

`setState` in dependency array is there because of eslint hooks rule. `state` is not there because in that example I don't use component state in `useEffect`.

We actually can and sometimes even have to use set function in `setState`, because sometimes it's the only way to get latest state and do something with it. `setState` in Recoil works exactly like `setState` in plain React, so set function is allowed.

**Comment by mondaychen** [comment]
Thanks @jj-143. It's interesting it's behaving differently in different react versions... Might be hard to investigate.
@chybisov which react version are you using?

**Comment by chybisov** [comment]
@mondaychen 17.0.2

**Comment by cyphire** [comment]
I understand what you mean about using setState in the useEffect, but I don't see why you are using it this way... I understand you are showing it as an anonymous example... but why show it this way?  My example was more real world...  I am not criticizing at all! just don't understand... Also you are showing [setState] in the bottom of the useEffect, don't you want [state] there?  This tell useEffect when to run again (when anything in the array [] changes.  You are passing the setter function in.  It won't ever change correct?

**Comment by chybisov** [comment]
@cyphire purpose of my example was to show library authors minimal steps to reproduce the warning and real code just doesn't matter in this context. 😃

**Comment by AMSTKO** [comment]
I have the same problem

react-native-web 0.13
react-native sdk-41.0.0
react 16.13.1
expo 41.0.0
react-navigation 5

On both web and mobile I have this warning when I switch to another "page" with the same useRecoilValue, after setted a simple atom with useSetRecoilState from the first page

**Comment by cyphire** [comment]
Hey let me know when you used the set useSetRecoilState setter was it in a useEffect?

**Comment by AMSTKO** [comment]
Hey @cyphire ! I use useSetRecoilState in a function after an onPress by the user. Not in the useEffect hook. 
I can live with his erros for now and I'm crossing my fingers for a resolution with react 17 that will come within 3 months in Expo

**Comment by cyphire** [comment]
Share your code if you world like I'll take a quick look at it.

**Comment by cyphire** [comment]
'''

	useEffect(() => {

		let unmounted = false;

		if (!unmounted)
			xxxxxxxxxxxxxxxxx
		else 
			console.log(`did not call as unmounted 1`)
		} else {
			let savedObj = JSON.parse(result)
  
			if (!unmounted) {
				setStatusObj(savedObj, 4)
				setStatusFinished(true);
			} else 
				console.log(`did not call as unmounted 2`)
			}
		})
	}
		} else {
			if (!unmounted) {
				setSaveRating(statusObj.ms_rating);
				setStatusFinished(true);
			}
			else 
				console.log(`did not call as unmounted 3`)

		}
		return () => { unmounted = true }
	}, [statusFinished, statusObj, inits, localStatusUpdate])
'''

Removed some of the code but note the
```	return () => { unmounted = true }```
and
``` if (!unmounted) {```

stopped the useEffect problem in my code

**Comment by AMSTKO** [comment]
I made this demo but without errors 

https://snack.expo.io/@amstko/react-native-recoil

- Go to screen B
- Set All screen titles in red
- Back to screen A
- Set all screen titles in red 

In this example: no problem.
In my stack with a more complexe code: "Can't perform a React state update on an unmounted component..."
He try to update value from useRecoilValue on the unmounted screen

I'll investigate my code for see differences but I have 3 sub level of navigation with a lot of components.

**Comment by cyphire** [comment]
So in your tree you definitely have a useEffect firing...  (your error shows it).  So the problem is somwhere up the tree. 
See if you comment out where you THINK the problem is see if you still have the same problem.. it might be higher up the tree...

**Comment by GimpMaster** [comment]
Just wanted to comment that I am seeing this issue too. I have a custom page router. I reduced my page down to just the useRecoilState with a return null.

When I switch to that page (which is now empty) and then to a different page and make a change to the same atom used on both pages I'll get errors that were mentioned above.

I'm currently using React "16.13.1"

This just started happening after upgrading to Recoil 0.2.0

**Comment by GimpMaster** [comment]
Ok, I wrote a real quick Code Sandbox that illustrates the problem.

Here is the link to the Code Sandbox

https://codesandbox.io/s/recoil-breaking-hxmnj?file=/src/App.js

Just incase the link doesn't work here is the main App.js with the problem.

```Javascript
import "./styles.css";
import { useState } from "react";

import { atom, useRecoilState, useRecoilValue } from "recoil";

const textState = atom({
  key: "textState", // unique ID (with respect to other atoms/selectors)
  default: "" // default value (aka initial value)
});

function Page1() {
  const [text, setText] = useRecoilState(textState);
  return <div onClick={() => setText("Page1")}>Page1</div>;
}

function Page2() {
  const [text, setText] = useRecoilState(textState);
  return <div onClick={() => setText("Page2")}>Page2</div>;
}

export default function App() {
  const [page, setPage] = useState(1);

  return (
    <div className="App">
      {page === 1 && <Page1 />}
      {page === 2 && <Page2 />}
      <h3
        onClick={() => {
          console.log("here");
          page === 1 ? setPage(2) : setPage(1);
        }}
      >
        Click To Change Page
      </h3>
    </div>
  );
}

```

To get the issue to appear. Click on "Click To Change Page". Then click on the "Page 2" div. Since the recoil State was already on Page 1, but now it unmounted it will still fire and cause the issue.

`Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.`

**Comment by davwheat** [comment]
I can reproduce with React Native 0.63, with React 16.13.1. I'm stuck on 0.63 because of Expo.

```tsx
import React from "react";
import { View } from "react-native";
import { TextArea } from "native-base";
import { useRecoilState } from "recoil";

const Header: React.FC<{ atom: Props["atom"] }> = ({ atom }) => {
  const [stationSelectFilter, setStationSelectFilter] = useRecoilState(atom);

  return (
    <View style={[styles.header]}>
      <TextArea
        style={styles.textField}
        placeholder="Search..."
        value={stationSelectFilter.filter}
        onChange={(e) => {
          // >> This is the issue <<
          setStationSelectFilter({
            filter: e.nativeEvent.text,
            selected: stationSelectFilter.selected,
          });
        }}
      />
    </View>
  );
};
```

In this example, I'm taking in an atom (this is part of a reusable component which is designed to be able to modify different states at will) and then updating it when text is entered.

----

I haven't been able to reproduce it in my other web-based Gatsby projects, which use React 17 in development and Preact in production.

**Comment by GimpMaster** [comment]
@davwheat - Yeah I really think its a regression Recoil has with React 16.X

**Comment by irinakk** [comment]
Repro this in react16 + recoil@0.2.0.
I traced a little and found that in my scenario its related to [`subscription.release` not called in this path](https://github.com/facebookexperimental/Recoil/blob/f17888e26d119250f06b7ffe6db833831dfc2654/src/hooks/Recoil_Hooks.js#L369) , but i dont know whether its designed this way.

**Comment by danieljaguiar** [comment]
Getting exact same issue... 

    "react": "16.14.0",
    "recoil": "0.2.0",

**Comment by Svarto** [comment]
Same issues started appearing for me when upgrading recoil:

    "react": "16.13.1",
    "recoil": "^0.2.0",

**Comment by ianstormtaylor** [comment]
Was having this and couldn't figure out why. Finally tracked it down to a component that needed to be wrapped in `<Suspense>` (because of a nested async atom) but was not. I don't know if there's a way to improve the error message when that happens on the Recoil/React end, but if so it would be greatly appreciated because it was a huge pain to debug.

**Comment by davwheat** [comment]
OH MY GOD THAT WOULD BE THE ISSUE!

Thank you!

**Comment by danieljaguiar** [comment]
I don't have any async atom but I still get the issue. @GimpMaster could repro the issue here without async: https://codesandbox.io/s/recoil-breaking-hxmnj?file=/src/App.js

**Comment by tabvn** [comment]
i got same issue.

ERROR
14:22
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

Stack trace:
  node_modules/react-native/Libraries/LogBox/LogBox.js:173:8 in registerError
  node_modules/react-native/Libraries/LogBox/LogBox.js:59:8 in errorImpl
  node_modules/react-native/Libraries/LogBox/LogBox.js:33:4 in console.error
  node_modules/expo/build/environment/react-native-logs.fx.js:27:4 in error
  node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:106:4 in printWarning
  node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:75:16 in error
  node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19152:9 in warnAboutUpdateOnUnmountedFiberInDEV
  node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:17093:40 in scheduleUpdateOnFiber
  node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:11003:16 in dispatchAction
  [native code]:null in dispatchAction
  node_modules/recoil/native/recoil.js:4310:26 in subscribeToRecoilValue$1$argument_2
  node_modules/recoil/native/recoil.js:3322:18 in sendEndOfBatchNotifications
  node_modules/recoil/native/recoil.js:3375:33 in Recoil_Queue.enqueueExecution$argument_1
  node_modules/recoil/native/recoil.js:1042:3 in enqueueExecution
  node_modules/recoil/native/recoil.js:3359:4 in

**Comment by tabvn** [comment]
i have to switching to Redux on mobile app (react-native, Expo).

**Comment by victoriafrench** [comment]
Having the same issue with the latest version (`0.3.1`). If I have an `atom` that contains an array and a `selector` subscribing to it via `get`, if I add or remove anything from the list `atom`, the error is present on the first action. Subsequent actions do not present the error. If the `selector` is removed, the error goes away. Perhaps this is an issue with `get` not unsubscribing correctly? (These are simple objects, no async)

```js
const simpleListAtom = atom({
     key: 'simpleListAtom',
     default: [],
});

const selectedIdAtom = atom({
     key: 'selectedIdAtom',
     default: null,
});

const activeItemAtom = atom({
     key: 'activeItemAtom',
     get: ({ get }) => {
          const selectedId = get(selectedIdAtom);
          const list = get(simpleListAtom);
          if (!selectedId) return null;
          return list.find(o => (o.id === selectedId));
     },
});
```

This would produce the error when a new selectedId is chosen, once a delete or add has occurred.

```js
const selectedId = useRecoildValue(selectedIdAtom);
const setList = useSetRecoildValue(activeItemAtom);

const handleOnClick = useCallback((event) => {
     if (event.preventDefault) event.preventDefault();
     if (!selectedId) return;
     setList((state) => {
          const index = state.findIndex(o => (o.id === selectedId));
          if (index === -1) return state; // already deleted
          const newState = [...state];
          newState.splice(index, 1);
          return newState;
     });
}, [selectedId, setList]);
```

whomever is listening to `activeItemAtom` when `selectedIdAtom` changes (no non-null) after any call to the setter of `activeItemAtom` consistently has an issue.

Hope this helps.

**Comment by Svarto** [comment]
Reading the thread there seems to be two separate - but potentially related issues. 

1. Atoms are not unsubscribed properly anymore since 0.2.0 and React 16 when unmounting components (e.g. screens)
2. Async selectors also show issues where a potential fix could be to wrap them correctly in Suspense.

Would be helpful if someone with more knowledge than me helped in summarize as this issue is quickly growing in size with several examples and sandboxes that show the issue in different setting. 

Seems like a huge problem potentially, Hope there's a way to fix this...

**Comment by chybisov** [comment]
Having the same issue with the latest **0.4.0** version. 😞

**Comment by Svarto** [comment]
Same here. @chybisov are you on React 17 or 16? I'm still on React 16 due to Expo

**Comment by victoriafrench** [comment]
@Svarto React 16 DOM here

**Comment by chybisov** [comment]
@Svarto exact same situation. React 16 and Expo

**Comment by fxwx23** [comment]
Having same issue here using `"react": "16.13.1"` , `"react-native": "~0.63.4"` and `"recoil": "^0.4.0"` , but now solved upgrading to `"react": "17.0.1"` and `"react-native": "0.64.0"` .

**Comment by drarmstr** [comment]
React has removed this warning, please see [this discussion about it](https://github.com/reactwg/react-18/discussions/82)
